config/configtls/configtls.go

```diff
@@ -104,14 +109,30 @@ func (c TLSSetting) loadTLSConfig() (*tls.Config, error) {
 		return nil, fmt.Errorf("for auth via TLS, either both certificate and key must be supplied, or neither")
 	}
 
-	var certificates []tls.Certificate
+	var getCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	var getClientCertificate func(*tls.CertificateRequestInfo) (*tls.Certificate, error)
 	if c.CertFile != "" && c.KeyFile != "" {
-		var tlsCert tls.Certificate
-		tlsCert, err = tls.LoadX509KeyPair(filepath.Clean(c.CertFile), filepath.Clean(c.KeyFile))
+		lastReload := time.Now()
+		var cert tls.Certificate
+		cert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile)
 		if err != nil {
 			return nil, fmt.Errorf("failed to load TLS cert and key: %w", err)
 		}
-		certificates = append(certificates, tlsCert)
+		getCert := func() (*tls.Certificate, error) {
+			now := time.Now()
+			if c.ReloadInterval != 0 && lastReload.Add(c.ReloadInterval).Before(now) {
+				var newCert tls.Certificate
+				newCert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile)
+				if err != nil {
+					return nil, err
+				}
+				cert = newCert
+				lastReload = now
+			}
+			return &cert, nil
+		}
+		getCertificate = func(chi *tls.ClientHelloInfo) (*tls.Certificate, error) { return getCert() }
+		getClientCertificate = func(cri *tls.CertificateRequestInfo) (*tls.Certificate, error) { return getCert() }
```

Comment:
For better "isolation/encapsulation" consider to create a new type/struct that offers "getCert" func. To better understand what properties are passed to the func/what changes, etc.

Reply:
Having a functional programming background, I like closures, but it's easy enough to refactor this way.

Reply:
I may understand that, but that's how we cannot easily see that "lastReload" is changed within a multi thread code without being protected, and there may be other variables as well. So I really believe that my suggestion is the right call here, especially when dealing with multi threaded code.